app/src/main/java/nerd/tuxmobil/fahrplan/congress/preferences/DefaultSettingsRepository.kt

```diff
@@ -0,0 +1,188 @@
+package nerd.tuxmobil.fahrplan.congress.preferences
+
+import android.content.Context
+import android.content.SharedPreferences
+import androidx.core.content.edit
+import androidx.preference.PreferenceManager
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.channels.trySendBlocking
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.distinctUntilChanged
+import nerd.tuxmobil.fahrplan.congress.R
+
+internal class DefaultSettingsRepository(
+    private val context: Context,
+) : SettingsRepository {
+
+    private val preferences: SharedPreferences = PreferenceManager.getDefaultSharedPreferences(context.applicationContext)
+
+    init {
+        PreferenceManager.setDefaultValues(context, R.xml.prefs, false)
+    }
+
+    private val userPreferencesFlow = callbackFlow {
+        trySendBlocking(getSettings())
+
+        val listener = SharedPreferences.OnSharedPreferenceChangeListener { _, _ ->
+            trySendBlocking(getSettings())
+        }
+
+        preferences.registerOnSharedPreferenceChangeListener(listener)
+
+        awaitClose {
+            preferences.unregisterOnSharedPreferenceChangeListener(listener)
+        }
+    }.distinctUntilChanged()
+
+    override fun getSettingsStream(): Flow<Settings> {
+        return userPreferencesFlow
+    }
```

Comment:
I feel this should be a property instead:

``` kotlin
// SettingsRepository
val settings: Flow<Settings>
```

``` kotlin
// DefaultSettingsRepository
override val settings = userPreferencesFlow
```

Reply:
To me settings aren't really a property of `SettingsRepository` (like e.g. name is a property of a person). `SettingsRepository` is an entity used to retrieve settings. That's why I think a function is more appropriate.

But I don't care too much. I'll change it to a property if you prefer it that way. However, I recommend using either "stream" (describes the type without being too specific) or "flow" as a suffix. That way it's clear the type is not a single `Settings` instance. It'll also make things simpler if at a later time a way is needed to retrieve the current `Settings` instance.

Reply:
I prefer to use a property named `settings`. Its type already exposed that it can be observed. Encoding the type in its name makes future refactoring unnecessary complex / larger diff (e.g. in the past a migration from `settingsLiveData` to `settingsFlow`).

The analogy of being a property of a repository kind of clicks for me as in: a repository is a gateway / date source for some property. Hence, the property is "hosting" this property.

Reply:
OK, let's go with `settingsStream` then.